### PR TITLE
Issue #9271: Remove detailAstImp from Rcurly

### PR DESCRIPTION
**Pull request by AkMo3**
Closes: #9271 

Continutation of #9288

Interestingly according to GitHub  [renaming](https://github.com/github/renaming/) the PR should automatically get targeted on renaming branch. But didn't happen with this one. It might be targeting PR on this branch and not from this branch. 

Sorry for that. Fixed points as pointed in #9288 in recent commit.

Diff Regression config: https://raw.githubusercontent.com/AkMo3/akmo3.github.io/Issue9217/Issue9271/config-Issue9271-20210224194247.xml
Diff Regression projects: https://raw.githubusercontent.com/AkMo3/akmo3.github.io/Issue9217/Issue9271/projects-to-test-on.properties

### CODE REVIEW

**Comment by AkMo3** [comment]
Github, generate report

**Comment by strkkk** [comment]
@AkMo3 report shows that there is some diff between existing implementation and your code.

This issue is supposed to be a refactoring, so no diff is expected. Please take a look why some violations disappeared.

**Comment by AkMo3** [comment]
GitHub, generate report

**Comment by AkMo3** [comment]
@strkkk Aren't these violations correct in the recent change, and they were not covered in the previous check.

**Comment by AkMo3** [comment]
Github, generate report

**Comment by AkMo3** [comment]
@strkkk Changes done to fix diff. Please review.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 1]
```diff
@@ -454,19 +459,28 @@ private static boolean skipDoubleBraceInstInit(Details details) {
...
+            final DetailAST rcurly = details.rcurly;
+            final DetailAST lcurly = details.lcurly;
+            isBlockAloneOnSingleLine = TokenUtil.areOnSameLine(rcurly, lcurly)
                 && (!TokenUtil.areOnSameLine(rcurly, nextToken)
                 || isRightcurlyFollowedBySemicolon(details));
+        }
+        if (nextToken == null
+                && TokenUtil.areOnSameLine(details.lcurly, details.rcurly)) {
+            isBlockAloneOnSingleLine = true;
+        }
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:482]: this 2 ifs can be simplified to something like
```
return TokenUtil.areOnSameLine(details.lcurly, details.rcurly) && (nextToken == null || !TokenUtil.areOnSameLine(rcurly, nextToken)
                || isRightcurlyFollowedBySemicolon(details))
```

**Review by strkkk — CHANGES_REQUESTED** [review]
last item

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 1 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done.

**Comment by strkkk** [comment]
Github, generate report

**Review by strkkk — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 2]
```diff
@@ -417,8 +415,10 @@ private static boolean shouldBeAloneOnLineWithNotAloneOption(Details details,
     private static boolean isAloneOnLine(Details details, String targetSrcLine) {
         final DetailAST rcurly = details.rcurly;
         final DetailAST nextToken = details.nextToken;
-        return (!TokenUtil.areOnSameLine(rcurly, nextToken) || skipDoubleBraceInstInit(details))
-                && CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(), targetSrcLine);
+        return (nextToken == null || !TokenUtil.areOnSameLine(rcurly, nextToken)
+                || skipDoubleBraceInstInit(details))
+                && CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(),
+                targetSrcLine);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:421]: Indentation looks weird. It looks like this line should be indented 4 spaces.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 3]
```diff
@@ -454,18 +459,18 @@ private static boolean skipDoubleBraceInstInit(Details details) {
      * @return true if block has single-line format and is alone on a line.
      */
     private static boolean isBlockAloneOnSingleLine(Details details) {
-        final DetailAST rcurly = details.rcurly;
-        final DetailAST lcurly = details.lcurly;
         DetailAST nextToken = details.nextToken;
-        while (nextToken.getType() == TokenTypes.LITERAL_ELSE) {
+        while (nextToken != null && nextToken.getType() == TokenTypes.LITERAL_ELSE) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:463]: Can the next token ever become `null` or is it always the first instance that is `null`?
Basically asking if the code should be switched to:
````
if (nextToken != null) {
    while (...) {
        ...
    }
}
````

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 4]
```diff
@@ -417,8 +415,10 @@ private static boolean shouldBeAloneOnLineWithNotAloneOption(Details details,
     private static boolean isAloneOnLine(Details details, String targetSrcLine) {
         final DetailAST rcurly = details.rcurly;
         final DetailAST nextToken = details.nextToken;
-        return (!TokenUtil.areOnSameLine(rcurly, nextToken) || skipDoubleBraceInstInit(details))
-                && CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(), targetSrcLine);
+        return (nextToken == null || !TokenUtil.areOnSameLine(rcurly, nextToken)
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:418]: So the way this reads, a `null` next token is always "alone on the line".

Can you show us a small example where one of these `null`s comes in?

Does a `null` next token always end up with no violations?

I ask to see if someone has checked if it makes more sense we add a global next token can't be `null` to the `validate` method  at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L325 instead of all these little areas.

**Comment by AkMo3** [comment]
Github, generate report.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 2 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 3 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Yeah, it can be null only at first instance since, after `if else` there would be atleast  another `rcurly` for end of class or interface. Thanks for pointing out.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 4 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:418]: > So the way this reads, a null next token is always "alone on the line".

Sir, what `null` only tells us that there is no token after it means it is the last of `file`, now the next part to check if there is there is something before it is handled by `CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(), targetSrcLine)`
and both are associated with and `&&` statement so both needs to be true. In short, no , if nextToken is `null` then it doesn't mean that token is alone.

> Can you show us a small example where one of these nulls comes in?


`public class InputRightCurlyEmptyOnSingleLine {} // violation`
-------------------------------------------------------------------------^----------

For this token, `nextToken` is `null`, since no next token is there, but it is not alone since `hasWhitespaceBefore()` is not statisfies.

> I ask to see if someone has checked if it makes more sense we add a global next token can't be null to the validate method at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L325 instead of all these little areas.

I guess it would be tricky as well since we need to count multiple scenarios of `nextToken` being null. Also, future changes being easy for specific condition with this setup :)

**Comment by AkMo3** [comment]
Github, generate report

**Comment by AkMo3** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 4 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:418]: Ok, from all the NPEs that were gotten here in regression, I thought there was more going on then a simple class file's end right curly.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 5]
```diff
@@ -454,19 +459,21 @@ private static boolean skipDoubleBraceInstInit(Details details) {
...
-        while (nextToken.getType() == TokenTypes.LITERAL_ELSE) {
-            nextToken = Details.getNextToken(nextToken);
+        if (nextToken != null) {
+            while (nextToken.getType() == TokenTypes.LITERAL_ELSE) {
+                nextToken = Details.getNextToken(nextToken);
+            }
         }
-        if (nextToken.getType() == TokenTypes.DO_WHILE) {
+
+        if (nextToken != null && nextToken.getType() == TokenTypes.DO_WHILE) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:469]: Both this if statement and the previous one can be joined since they both have the same null check on the same variable.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 6]
```diff
@@ -417,8 +415,10 @@ private static boolean shouldBeAloneOnLineWithNotAloneOption(Details details,
     private static boolean isAloneOnLine(Details details, String targetSrcLine) {
         final DetailAST rcurly = details.rcurly;
         final DetailAST nextToken = details.nextToken;
-        return (!TokenUtil.areOnSameLine(rcurly, nextToken) || skipDoubleBraceInstInit(details))
-                && CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(), targetSrcLine);
+        return (nextToken == null || !TokenUtil.areOnSameLine(rcurly, nextToken)
+            || skipDoubleBraceInstInit(details))
+            && CommonUtil.hasWhitespaceBefore(details.rcurly.getColumnNo(),
+            targetSrcLine);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:421]: Like previously mentioned, this line should be indented.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 7]
```diff
@@ -454,19 +459,21 @@ private static boolean skipDoubleBraceInstInit(Details details) {
...
             final DetailAST doWhileSemi = nextToken.getParent().getLastChild();
             nextToken = Details.getNextToken(doWhileSemi);
         }
-        return TokenUtil.areOnSameLine(rcurly, lcurly)
-                && (!TokenUtil.areOnSameLine(rcurly, nextToken)
-                || isRightcurlyFollowedBySemicolon(details));
+
+        return TokenUtil.areOnSameLine(details.lcurly, details.rcurly)
+            && (nextToken == null || !TokenUtil.areOnSameLine(details.rcurly, nextToken)
+            || isRightcurlyFollowedBySemicolon(details));
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java:476]: same here since this is a continuation of the `(`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 5 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 6 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 7 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done

**Comment by AkMo3** [comment]
Github, generate report

File: config/import-control.xml [hunk 8]
```diff
@@ -14,9 +14,6 @@
   <!-- guava causes issues with other libraries since not everyone is on the same version -->
   <disallow pkg="com\.google.*" regex="true"/>
 
-  <!-- until https://github.com/checkstyle/checkstyle/issues/9271 -->
```
Comment (AkMo3) [inline config/import-control.xml:17]: @strkkk @rnveach We had to undo this change, since ParserUtil is using this.

**Review by AkMo3 — COMMENTED** [review]
Please see:

File: config/import-control.xml [hunk 9]
```diff
@@ -60,6 +57,7 @@
   </file>
   <file name="ParserUtil">
     <allow class="antlr.Token"/>
+    <allow class="com.puppycrawl.tools.checkstyle.DetailAstImpl"/>
```
Comment (AkMo3) [inline config/import-control.xml:60]: @strkkk @rnveach Added this line to allow DetailAstImpl being used here.

File: config/import-control.xml (hunk 8 above)
Comment (AkMo3) [reply config/import-control.xml:17]: No more changes required.

File: config/import-control.xml (hunk 9 above)
Comment (rnveach) [reply config/import-control.xml:60]: This is only because you removed the global and added a finer restriction to just the class.
This is fine to me.

**Review by rnveach — APPROVED** [review]

**Review by romani — APPROVED** [review]
